snowfall/training/compute_expected_times.py

```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env python3
+#
+# Copyright (c)  2021  Xiaomi Corp.       (author: Fangjun Kuang)
+#
+
+import k2
+import torch
+
+
+def _create_phone_fsas(phone_seqs: k2.RaggedInt) -> k2.Fsa:
+    '''
+    Args:
+      phone_seqs:
+        It contains two axes with elements being phone IDs.
+        The last element of each sub-list is -1.
+    Returns:
+      Return an FsaVec representing the phone seqs.
+    '''
+    assert phone_seqs.num_axes() == 2
+    phone_seqs = k2.ragged.remove_values_eq(phone_seqs, -1)
+    return k2.linear_fsa(phone_seqs)
+
+
+def compute_expected_times_per_phone(mbr_lats: k2.Fsa,
+                                     ctc_topo: k2.Fsa,
+                                     dense_fsa_vec: k2.DenseFsaVec,
+                                     use_double_scores=True,
+                                     num_paths=100) -> torch.Tensor:
+    '''Compute expected times per phone in a n-best list.
+
+    See the following comments for more information:
+
+        - `<https://github.com/k2-fsa/snowfall/issues/96>`_
+        - `<https://github.com/k2-fsa/k2/issues/641>`_
+
+    Args:
+      mbr_lats:
+        An FsaVec.
+      ctc_topo:
+        The return value of :func:`build_ctc_topo`.
+      dense_fsa_vec:
+        It contains nnet_output.
+      use_double_scores:
+        True to use `double` in :func:`k2.random_paths`; false to use `float`.
+      num_paths:
+        Number of random paths to draw in :func:`k2.random_paths`.
+    Returns:
+      A 1-D torch.Tensor contains the expected times per pathphone_idx.
+    '''
+    lats = mbr_lats
+    assert len(lats.shape) == 3
+    assert hasattr(lats, 'phones')
+
+    # paths will be k2.RaggedInt with 3 axes: [seq][path][arc_pos],
+    # containing arc_idx012
+    paths = k2.random_paths(lats,
+                            use_double_scores=use_double_scores,
+                            num_paths=num_paths)
+
+    # phone_seqs will be k2.RaggedInt like paths, but containing phones
+    # (and final -1's, and 0's for epsilon)
+    phone_seqs = k2.index(lats.phones, paths)
+
+    # Remove epsilons from `phone_seqs`
+    print('before removing 0', phone_seqs.shape().row_splits(2))
```

Comment:
It shows that phones_seqs from mbr_lats contain more 0s.


<img width="1021" alt="Screen Shot 2021-02-23 at 4 14 40 PM" src="https://user-images.githubusercontent.com/5284924/108817244-8ba03580-75f2-11eb-87f3-2c6d8c2cf053.png">

<img width="1024" alt="Screen Shot 2021-02-23 at 4 14 54 PM" src="https://user-images.githubusercontent.com/5284924/108817251-8e028f80-75f2-11eb-9487-07cee29970dd.png">

Reply:
can you verify that lats.phones is a Tensor and not a _k2.RaggedInt, in both cases?

Reply:
Yes, I confirm that they are both 1-D tensors.